python/dolfinx/fem/forms.py

```diff
@@ -199,29 +200,52 @@ def _form(form):
         ]
         constants = [c._cpp_object for c in form.constants()]
 
+        # Make map from integral_type to the meshtag indices of interest
+        subdomain_ids = {type: [] for type in sd.get(domain).keys()}
+        for integral in form.integrals():
+            if integral.subdomain_data() is not None:
+                subdomain_ids[integral.integral_type()].append(integral.subdomain_id())
+        subdomain_ids = {key: list(chain(values)) for key, values in subdomain_ids.items()}
+
         # NOTE Could remove this and let the user convert meshtags by
         # calling compute_integration_domains themselves
-        def get_integration_domains(integral_type, subdomain):
+        def get_integration_domains(integral_type, subdomain, subdomain_ids):
             """Get integration domains from subdomain data"""
             if subdomain is None:
                 return []
             else:
+                domains = []
                 try:
                     if integral_type in (IntegralType.exterior_facet, IntegralType.interior_facet):
                         tdim = subdomain.topology.dim
                         subdomain._cpp_object.topology.create_connectivity(tdim - 1, tdim)
                         subdomain._cpp_object.topology.create_connectivity(tdim, tdim - 1)
-                    domains = _cpp.fem.compute_integration_domains(
-                        integral_type, subdomain._cpp_object
-                    )
+                    # Compute integration domains only for each subdomain id in the integrals
+                    # If a process has no integral entities, insert an empty array
+                    for id in subdomain_ids:
+                        try:
+                            integration_entities = _cpp.fem.compute_integration_domains(
```

Comment:
It seems a bit strange to loop over the IDs in Python and call `compute_integration_entities` on each when `compute_integration_domains` loops over IDs anyway. Maybe the fix should happen inside `compute_integration_domains` instead so that it returns `[..., (id, []), ....]` when there are no entities to integrate over for that ID? That way, we can just call it once and not have to pass `subdomain.values[subdomain.values == id]`, which is just a (possibly quite large) array filled with the integral ID. If there's a reason we can't do this, then maybe we should change `compute_integration_domains` so that it only does one ID at a time, but I think the first option is better.

Reply:
`compute_integration_domains` only loops over ids present on the current process. It has no notion of what is on the other processes, or what is in the integral form. We cannot loop over every possible id (as it is an int 32).
Preferably, I would change `compute_integration_entities` to only take in the id of the entities, i.e.
```c++
/// @param[in] integral_type Integral type
/// @param[in] topology Mesh topology
/// @param[in] entities List of tagged mesh entities
/// @param[in] dim Topological dimension of tagged entities
/// @return List of `(integral id, entities)` pairs
/// @pre The topological dimension of the integral entity type and the
/// topological dimension of mesh tag data must be equal.
/// @pre For facet integrals, the topology facet-to-cell and
/// cell-to-facet connectivity must be computed before calling this
/// function.
std::vector<std::int32_t>
compute_integration_domains(IntegralType integral_type,
                            const mesh::Topology& topology,
                            std::span<const std::int32_t> entities, int dim);
```
then the user can associate what-ever value with the integration entities.